Add web push architecture guide (#43419)
mdn/content: files/en-us/web/api/push_api/web_push_architecture/index.md

---

> title: Web push architecture

Reviewer: ```suggestion
title: Web Push architecture
```

(plus any other changes needed for this capitalization fix). Everywhere else we capitalize "Web Push".

slug: Web/API/Push_API/Web_Push_Architecture
page-type: guide
---

{{DefaultAPISidebar("Push API")}}

This article explains how the different pieces of the Web Push infrastructure fit together, from subscribing to a push service through to delivering a push message to a service worker.

## Overview

The Web Push system involves several parties working together:

> 1. Your **web application**, running in the user's browser

Reviewer: I don't think the bold formatting is needed here.

> 2. A **service worker**, registered by your application to receive push events in the background

Reviewer: The service worker is part of the web application: from an architecture point of view, this is one piece. We could replace 1 and 2 with something like: "A service worker, which is part of the web application, and which subscribes to push events and handles them when the browser delivers them to the web application."

3. A **push service**, operated by the browser vendor (such as Mozilla or Google), which routes messages from your server to the right browser
4. Your **application server**, which sends messages to the push service when it needs to notify a user

> 

Reviewer: The browser is a piece that's missing here.

The key insight is that your server never talks directly to the user's browser. Instead, it posts messages to a push service, which holds them until the browser comes online to collect them.

> 

Reviewer: I think it would be better not to have a preamble sentence and not to have an H2 Overview, but instead have a couple of paragraphs giving a very high-level outline of the WP architecture, which itself provides the motivation for this document (i.e. the point that in WP a browser-specific service mediates the communication between the web server and the web app).

Reviewer: I think it's also worth including a diagram, and some indication of the implications of this architecture, especially as it relates to signing and encryption. A lot of the complexity comes from signing and encryption, and without a higher-level view of the motivation for that, it seems to come out of nowhere.

There's an overview of web push in the PWA docs: https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/Guides/Offline_and_background_operation#push, that includes a diagram and some overview text:

> Push messages are not sent directly from the app server to the device. Instead, your app server sends messages to a push service, from which the device can retrieve them and deliver them to the app.
>
> This also means that messages from your server to the push service need to be [encrypted](https://developer.mozilla.org/en-US/docs/Glossary/Encryption) (so the push service can't read them) and [signed](https://developer.mozilla.org/en-US/docs/Glossary/Signature/Security) (so the push service knows that the messages are really from your server, and not from someone impersonating your server).
>
> The push service is operated by the browser vendor or by a third party, and the app server communicates with it using the [HTTP Push](https://datatracker.ietf.org/doc/html/rfc8030) protocol. The app server can use a third-party library such as [web-push](https://github.com/web-push-libs/web-push) to take care of the protocol details.

I think it would be worth integrating this content into the new page.

> ## The subscription flow

Reviewer: This omits server provisioning, which happens before subscription (and without which, we don't know where "your server's VAPID public key" comes from).

> ## The subscription flow

Reviewer: ```suggestion
## Subscription
```

> Before your server can send push messages, the user's browser must subscribe to the push service:

Reviewer: It's not the browser that subscribes. It's your web application.

1. Your application requests notification permission from the user.

> 2. On approval, it calls {{domxref("PushManager.subscribe()")}} with your server's VAPID public key in the `applicationServerKey` option.

Reviewer: I'm not sure if the user always has to grant permission (or if it does every time, as this implies)

Reviewer: This mentions "your server's VAPID public key" but we haven't talked about VAPID yet. It would be better perhaps to say something like "The application server's public key, which the push service will use to verify the signature on push messages from the application server."

3. The browser contacts its push service and creates a new subscription.
4. {{domxref("PushManager.subscribe()")}} resolves with a {{domxref("PushSubscription")}} object containing:
   - An **endpoint** URL — a unique URL on the push service where your server will POST messages.
   - A **p256dh** key — the browser's public key for encrypting message payloads.

>    - An **auth** secret — a shared secret for message authentication.

Reviewer: I'm unclear what `auth` is for, it's not mentioned anywhere else.

>    - An **auth** secret — a shared secret for message authentication.

Reviewer: ```suggestion
   - `endpoint`: a unique URL on the push service where your server will POST messages.
   - `p256dh`: the browser's public key, which the application server can use to encrypt push messages.
   - `auth`: a shared secret for message authentication.
```

The identifiers here should all link to the relevant bit of the reference docs, e.g. https://developer.mozilla.org/en-US/docs/Web/API/PushSubscription/getKey#p256dh).

5. Your application sends this subscription data to your server, which stores it.

> The endpoint URL is a [capability URL](https://w3ctag.github.io/capability-urls/): anyone who knows it can send messages to that browser. Your server must keep it secret.

Reviewer: "anyone who knows it can send messages to that browser" this is surely not true, because the messages are also signed.

## Push services

Each browser vendor operates its own push service:

- **Firefox** uses [Mozilla Autopush](https://mozilla-push-service.readthedocs.io/), a Python/Rust service that maintains WebSocket connections to Firefox clients and exposes REST endpoints for application servers.
- **Chrome and Chromium-based browsers** use [Firebase Cloud Messaging (FCM)](https://firebase.google.com/docs/cloud-messaging), which handles routing through Google's infrastructure.
- **Safari on macOS and iOS** uses Apple's push infrastructure. On iOS and iPadOS (16.4+), web push follows the standard W3C Push API but is only available for Progressive Web Apps installed to the home screen.

> Despite the different backend implementations, all push services speak the same protocol: the [Web Push Protocol (RFC 8030)](https://datatracker.ietf.org/doc/html/rfc8030). This means your application server uses the same code to send messages regardless of which browser the user has. The endpoint URL in each subscription already points to the correct push service.

Reviewer: I don't think this section is needed at all. It's transparent to web devs (and must be: browser vendors must be able to change these arrangements without breaking web apps).

## Sending a message

> When your server needs to notify a user, it sends an HTTP POST request to the subscription's endpoint URL. The request includes:

Reviewer: ```suggestion
When your server needs to notify a user, it sends an HTTP {{httpmethod("POST")}} request to the subscription's endpoint URL. The request includes:
```

Similar comment applies in other places: cross-reference to MDN wherever it's appropriate. This includes specific parts of the web platform like HTTP headers and methods, but also glossary entries for concepts like encryption or signing.

Reviewer: When I see "The request includes:" I expect to see next a list of specific items in the request. But here the next thing is a heading, which generally serves as a structural component in a page. Also, not all the following headings describe things that are in the request: for example:

- "VAPID authentication" is not itself a thing, it's a process
- "Response codes" is not a thing in the request (obviously)

So this needs some rework.

### Encrypted payload

Message content is encrypted end-to-end so the push service cannot read it. The encryption uses:

- **ECDH key agreement** (P-256) between your server and the browser's subscription keys
- **HKDF-SHA-256** for deriving encryption keys

> - **AES-128-GCM** content encoding (defined in [RFC 8188](https://datatracker.ietf.org/doc/html/rfc8188))

Reviewer: AES-128-GCM is not content encoding. Also I think just listing acronyms isn't terribly helpful here. You could talk about how these ciphers are used together to encrypt the message.

> The full protocol is specified in [RFC 8291 (Message Encryption for Web Push)](https://datatracker.ietf.org/doc/html/rfc8291). In practice, libraries like [web-push](https://github.com/web-push-libs) handle this automatically.

Reviewer: It's odd to mention web-push here specifically, since AFAIK web-push handles everything that the app server needs, including things like VAPID.
